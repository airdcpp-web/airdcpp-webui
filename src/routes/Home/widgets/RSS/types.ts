import * as UI from '@/types/ui';

export type FeedNode<Attributes extends object = UI.EmptyObject> =
  | string
  | {
      text: string;
      attr: Attributes;
    };

export interface FeedItem {
  title?: FeedNode;
  updated?: FeedNode;
  pubDate?: FeedNode;
  guid?: FeedNode;
  id?: FeedNode;
  link?: FeedNode<{
    href: string;
  }>;
}

export interface RawFeedData {
  rss?: {
    channel?: {
      item?: FeedItem[];
    };
  };
  feed?: {
    entry?: FeedItem[];
  };
}
