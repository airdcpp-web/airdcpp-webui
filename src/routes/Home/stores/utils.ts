import { saveWidgetSettings } from '@/routes/Home/widgets/WidgetUtils';
import { Layouts } from 'react-grid-layout';

import * as UI from '@/types/ui';
import {
  ApplicationWidgetInfo,
  ExtensionWidgetInfo,
  NotepadWidgetInfo,
  RSSWidgetInfo,
  TransferWidgetInfo,
} from '../widgets';
import { HomeLayoutColumns } from '../constants/HomeLayoutConstants';

export const createLayoutWidget = (
  layouts: Layouts,
  widgetInfo: UI.Widget,
  id: string,
  x?: number,
  y?: number,
) => {
  // Add the same widget for all layouts (we are lazy and use the same size for all layouts)
  return Object.keys(HomeLayoutColumns).reduce((reducedLayouts, key) => {
    reducedLayouts[key] = layouts[key] || [];
    reducedLayouts[key] = reducedLayouts[key].concat({
      i: id,
      x: x || (reducedLayouts[key].length * 2) % (HomeLayoutColumns[key] || 12),
      y: y || 0,
      ...widgetInfo.size,
    });

    return reducedLayouts;
  }, {} as Layouts);
};

const createDefaultLayoutWidget = <SettingsT>(
  layouts: Layouts,
  widgetInfo: UI.Widget,
  x: number,
  y: number,
  name?: string,
  settings?: SettingsT,
  suffix = '_default',
) => {
  const id = widgetInfo.typeId + suffix;
  const newLayout = createLayoutWidget(layouts, widgetInfo, id, x, y);

  saveWidgetSettings(id, {
    name,
    widget: settings,
  });

  return newLayout;
};

const createDefaultLayoutWidgets = (layouts: Layouts) => {
  layouts = createDefaultLayoutWidget(layouts, ApplicationWidgetInfo, 0, 0);
  layouts = createDefaultLayoutWidget(
    layouts,
    RSSWidgetInfo,
    ApplicationWidgetInfo.size.w,
    0,
    'News',
    {
      feed_url: 'https://airdcpp-web.github.io/feed.xml',
    },
    '_releases',
  );

  layouts = createDefaultLayoutWidget(
    layouts,
    ExtensionWidgetInfo,
    ApplicationWidgetInfo.size.w + RSSWidgetInfo.size.w,
    0,
  );
  layouts = createDefaultLayoutWidget(
    layouts,
    TransferWidgetInfo,
    ApplicationWidgetInfo.size.w + RSSWidgetInfo.size.w + ExtensionWidgetInfo.size.w,
    0,
  );
  layouts = createDefaultLayoutWidget(layouts, NotepadWidgetInfo, 0, 5);
  return layouts;
};

export const ensureDefaultLayoutWidgets = (layouts: Layouts) => {
  if (Object.keys(layouts).length !== 0) {
    return layouts;
  }

  // Initialize the default layout
  return createDefaultLayoutWidgets(layouts);
};
