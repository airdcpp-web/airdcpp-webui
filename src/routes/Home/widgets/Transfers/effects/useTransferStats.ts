import { useEffect, useState } from 'react';

import TransferConstants from '@/constants/TransferConstants';

import * as API from '@/types/api';

import { AddSocketListener } from '@/decorators/SocketSubscriptionDecorator';

import { APISocket } from '@/services/SocketService';

const IDLE_APPEND_INTERVAL_MS = 3000;
const DEFAULT_MAX_EVENTS = 1800; // 30 minutes when transfers are running

type Points = Array<number[]>;

interface Series {
  points: Points;
  maxDownload: number;
  maxUpload: number;
}

interface State {
  series: Series;
  stats: API.TransferStats | undefined;
}

const addSpeed = (
  { points, maxDownload: prevMaxDownload, maxUpload: prevMaxUpload }: Series,
  down: number,
  up: number,
  maxEvents: number = DEFAULT_MAX_EVENTS,
) => {
  const ret = [...points, [Date.now(), down, up]];

  if (ret.length > maxEvents) {
    ret.shift();
  }

  return {
    points: ret,
    maxDownload: Math.max(down, prevMaxDownload),
    maxUpload: Math.max(up, prevMaxUpload),
  };
};

const hasEllapsedSinceLastUpdate = (points: Points, ms: number) => {
  return points[points.length - 1][0] + ms <= Date.now();
};

export const useTransferStats = (
  socket: APISocket,
  addSocketListener: AddSocketListener,
  maxEvents: number = DEFAULT_MAX_EVENTS,
) => {
  const [state, setState] = useState<State>({
    series: {
      points: [[Date.now(), 0, 0]],
      maxDownload: 0,
      maxUpload: 0,
    },
    stats: undefined,
  });

  const checkIdle = () => {
    setState((oldState) => {
      if (
        !hasEllapsedSinceLastUpdate(oldState.series.points, IDLE_APPEND_INTERVAL_MS - 200)
      ) {
        return oldState;
      }

      return {
        ...oldState,
        series: addSpeed(oldState.series, 0, 0, maxEvents),
      };
    });
  };

  const onStatsReceived = (newStats: Partial<API.TransferStats>) => {
    setState((prevState) => ({
      series: addSpeed(
        prevState.series,
        newStats.speed_down || prevState.stats?.speed_down || 0,
        newStats.speed_up || prevState.stats?.speed_up || 0,
        maxEvents,
      ),
      stats: {
        ...prevState.stats,
        ...(newStats as API.TransferStats),
      },
    }));
  };

  const onStatsUpdated = (newStats: Partial<API.TransferStats>) => {
    // Ignore updates that are received in bursts
    // The precision of Date.now() has been slightly reduced/randomized,
    // which is possibly causing crashes in the TimeSeries creation at the render method
    // due to unchronological events in the series
    //
    // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/now
    //if (!hasEllapsedSinceLastUpdate(points, 100)) {
    //  console.debug(`Transfer widget: burst update ignored`);
    //  return;
    //}

    onStatsReceived(newStats);
  };

  const initStats = async () => {
    try {
      const stats = await socket.get<API.TransferStats>(TransferConstants.STATISTICS_URL);
      onStatsReceived(stats);
    } catch (error) {
      console.error('Failed to fetch transfer statistics', error.message);
    }

    // Add lister
    addSocketListener(
      TransferConstants.MODULE_URL,
      TransferConstants.STATISTICS,
      onStatsUpdated,
    );
  };

  useEffect(() => {
    initStats();

    const idleInterval = window.setInterval(checkIdle, IDLE_APPEND_INTERVAL_MS);
    return () => {
      clearInterval(idleInterval);
    };
  }, []);

  return state;
};
