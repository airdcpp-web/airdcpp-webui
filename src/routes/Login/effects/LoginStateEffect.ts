import { useEffect, useState } from 'react';
import { LoginAPIActions } from '@/actions/store/LoginActions';

import { useTranslation } from 'react-i18next';
import { useLocation, useNavigate } from 'react-router';
import { useSocket } from '@/context/SocketContext';
import { useAppStoreApi, useAppStoreProperty } from '@/context/AppStoreContext';

export interface LoginLocationState {
  nextPath?: string;
}

const useSessionState = () => {
  const [loading, setLoading] = useState(false);
  const { i18n } = useTranslation();
  const location = useLocation();
  const navigate = useNavigate();
  const socket = useSocket();
  const appStoreApi = useAppStoreApi();

  const login = useAppStoreProperty((state) => state.login);

  const refreshToken = login.getRefreshToken();
  const session = login.getSession();

  const checkLoginState = () => {
    if (login.socketAuthenticated) {
      // Set UI language to match the app language
      const { system_info: systemInfo } = session!;
      if (systemInfo) {
        const language = systemInfo.language;
        if (language !== i18n.language) {
          i18n.changeLanguage(language);
        }
      }

      // Redirect to the main app
      const state = location.state as LoginLocationState;
      const nextPath = state?.nextPath ?? '/';
      navigate(nextPath, { replace: true });
    } else if (!!login.lastError && !refreshToken) {
      // Keep the loading state as true as long as there is socket actions happening
      // Changing the state to false will allow the login page to update

      // Failed
      setLoading(false);
    }
  };

  useEffect(() => {
    checkLoginState();
  }, [login]);

  useEffect(() => {
    if (!!refreshToken) {
      LoginAPIActions.loginRefreshToken(refreshToken, {
        socket,
        appStore: appStoreApi.getState(),
      });
      setLoading(true);
    }
  }, []);

  return { loading, setLoading };
};

export { useSessionState };
