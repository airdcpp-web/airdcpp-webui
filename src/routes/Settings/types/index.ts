import * as API from '@/types/api';
import * as UI from '@/types/ui';

export type SettingSectionLayoutProps = Pick<UI.RouteComponentProps, 'location'> &
  React.PropsWithChildren<{
    contentClassname: string;
    selectedRootMenuItem: RootSectionType;
    selectedChildMenuItem: ChildSectionType;
    menu: {
      childMenuItems: React.ReactNode[];
      childAdvancedMenuItems?: React.ReactNode[];
      rootMenuItems: React.ReactNode[];
    };
    settingsT: UI.ModuleTranslator;
    message: React.ReactNode;
    getSaveButton: (className?: string) => React.ReactNode;
  }>;

export type SettingPageProps = React.PropsWithChildren<{
  settingsT: UI.ModuleTranslator;
  moduleT: UI.ModuleTranslator;
}>;

export interface SectionBase {
  url: string;
  access?: API.AccessEnum;
  title: string;
}

export interface ChildSectionType extends SectionBase {
  noSave?: boolean;
  local?: boolean;
  component: React.ComponentType<SettingPageProps>;
}

export interface RootSectionType extends SectionBase {
  menuItems: ChildSectionType[];
  advancedMenuItems?: ChildSectionType[];
  icon: string;
  component: React.ComponentType;
}
