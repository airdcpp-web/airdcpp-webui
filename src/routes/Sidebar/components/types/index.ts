//import * as React from 'react';

import { IconType } from 'components/semantic/Icon';

import * as API from 'types/api';
import * as UI from 'types/ui';

export interface SessionLayoutLabels {
  // Label for button that opens a new session
  newCaption?: React.ReactNode;

  // Label for button that opens a new session
  newDescription?: React.ReactNode;

  // Label for button that opens a new session
  newIcon?: IconType;
}

export interface SessionLayoutManageProps<
  SessionT extends UI.SessionItemBase,
  SessionApiT extends object = UI.EmptyObject,
  UIActionT extends UI.ActionListType<UI.SessionItemBase> = UI.EmptyObject,
> {
  // Unique ID of the section (used for storing and loading the previously open tab)
  baseUrl: string;

  // Array of the items to list
  items: SessionT[];

  // Session actions to show in the action menu
  actionIds?: string[];

  uiActions: UI.ModuleActions<SessionT, void, UIActionT>;
  sessionApi: UI.SessionActions<SessionT> & SessionApiT;

  // Item ID that is currently active (if any)
  activeId: API.IdType | undefined;

  // AccessConstant defining whether the user has edit permission
  editAccess: API.AccessEnum;

  unreadInfoStore: UI.UnreadInfoStore;

  remoteMenuId: string;
}

interface SessionMainLayoutComponents {
  newButton: React.ReactElement<any> | null;
  listActionMenuGetter: () => React.ReactNode;
  itemHeaderTitle: React.ReactNode;
  itemHeaderDescription: React.ReactNode;
  itemHeaderIcon: IconType;
  sessionMenuItems: React.ReactNode[];
}

export interface SessionMainLayoutProps<
  SessionT extends UI.SessionItemBase,
  SessionApiT extends object = UI.EmptyObject,
  UIActionsT extends UI.ActionListType<UI.SessionItemBase> = UI.EmptyObject,
> extends Pick<
      SessionLayoutManageProps<SessionT, SessionApiT, UIActionsT>,
      'unreadInfoStore'
    >,
    SessionMainLayoutComponents {
  activeItem: SessionT | null;
  onKeyDown: (event: React.KeyboardEvent) => void;
  children: React.ReactNode;

  actions: SessionLayoutManageProps<SessionT, SessionApiT, UIActionsT>['uiActions'];
}

export type SessionChildProps<
  SessionT extends UI.SessionItemBase,
  SessionApiT extends object = UI.EmptyObject,
  UIActionsT extends UI.ActionListType<UI.SessionItemBase> = UI.EmptyObject,
> = Pick<
  SessionLayoutManageProps<SessionT, SessionApiT, UIActionsT>,
  'sessionApi' | 'uiActions'
> &
  UI.RouteComponentProps & {
    session: SessionT;
    sessionT: UI.ModuleTranslator;
  };

export interface NewSessionLayoutProps extends UI.RouteComponentProps {
  sessionT: UI.ModuleTranslator;
}

export interface SessionLocationState {
  pending?: boolean;
}
