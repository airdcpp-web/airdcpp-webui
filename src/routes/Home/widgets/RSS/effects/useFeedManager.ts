import { useEffect, useState } from 'react';

import { loadSessionProperty, saveSessionProperty } from '@/utils/BrowserUtils';

import { Settings } from '../';
import { FeedItem, RawFeedData } from '../types';
import { fetchRSSFeed, parseRSSFeed } from '../utils';

import * as UI from '@/types/ui';

import '../style.css';
import { useSession } from '@/context/AppStoreContext';

const idToCacheKey = (id: string) => `rss_feed_cache_${id}`;

interface StorageFeed {
  entries: FeedItem[];
  date: number;
}

// export type RSSProps = UI.WidgetProps<Settings>;

const getCachedFeedInfo = (componentId: string, settings: Settings) => {
  const feedInfo = loadSessionProperty<StorageFeed>(idToCacheKey(componentId));

  const { feed_url, feed_cache_minutes } = settings;
  if (feedInfo) {
    const feedDate = new Date(feedInfo.date).getTime();
    const lastValidDate = feedDate + feed_cache_minutes * 60 * 1000;

    if (lastValidDate >= Date.now()) {
      console.log(
        `RSS: cached feed will be used (expires in ${
          (lastValidDate - Date.now()) / 60 / 1000
        } minutes)`,
        feed_url,
      );

      return feedInfo;
    } else {
      console.log(
        `RSS: cached feed had expired ${
          (Date.now() - lastValidDate) / 60 / 1000
        } minutes ago`,
        feed_url,
      );
    }
  } else {
    console.log('RSS: no cached feed', feed_url);
  }

  return null;
};

export const useFeedManager = (props: UI.WidgetProps<Settings>) => {
  const { settings, widgetT, componentId } = props;

  const session = useSession();

  const [feedInfo, setFeedInfo] = useState<StorageFeed | null>(
    getCachedFeedInfo(componentId, settings),
  );
  const [error, setError] = useState<string | null>(null);

  /*componentDidMount() {
    if (!this.state.entries) {
      this.handleUpdate();
    }
  }*/

  const resetFeed = () => {
    if (feedInfo?.entries) {
      setFeedInfo(null);
    }

    if (error) {
      setError(null);
    }
  };

  const onFeedFetched = (data: RawFeedData) => {
    const entries = parseRSSFeed(data, widgetT);

    saveSessionProperty(idToCacheKey(componentId), {
      entries,
      date: Date.now(),
    });

    setFeedInfo({ entries, date: Date.now() });
  };

  const fetchFeed = async (feedUrl: string) => {
    if (!feedUrl.startsWith('http://') && !feedUrl.startsWith('https://')) {
      setError(widgetT.translate('Invalid URL'));
      return;
    }

    resetFeed();

    try {
      const jsonFeed = await fetchRSSFeed(feedUrl, session);
      onFeedFetched(jsonFeed);
    } catch (e) {
      console.log('RSS feed download failed', feedUrl, e);
      setError(e.message);
    }
  };

  const handleUpdate = () => {
    fetchFeed(settings.feed_url);
  };

  useEffect(() => {
    if (!feedInfo?.entries) {
      handleUpdate();
    }
  }, []);

  useEffect(() => {
    fetchFeed(settings.feed_url);
  }, [settings.feed_url]);

  return { error, feedInfo, handleUpdate };
};
