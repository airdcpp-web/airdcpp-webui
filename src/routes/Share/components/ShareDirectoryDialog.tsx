import { useMemo, useRef } from 'react';
import RouteModal from 'components/semantic/RouteModal';

import ShareConstants from 'constants/ShareConstants';
import ShareRootConstants from 'constants/ShareRootConstants';
import IconConstants from 'constants/IconConstants';

import ModalRouteDecorator, {
  ModalRouteDecoratorChildProps,
} from 'decorators/ModalRouteDecorator';

import ShareProfileDecorator, {
  ShareProfileDecoratorChildProps,
} from 'decorators/ShareProfileDecorator';

import Message from 'components/semantic/Message';

import t from 'utils/tcomb-form';

import { getLastDirectory } from 'utils/FileUtils';

import Form, {
  FormSaveHandler,
  FormFieldSettingHandler,
  FormFieldChangeHandler,
} from 'components/form/Form';
import FilesystemConstants from 'constants/FilesystemConstants';
import { AutoSuggestField } from 'components/form/fields';

import '../style.css';

import * as API from 'types/api';
import * as UI from 'types/ui';

import { ShareRootEntryBase } from 'types/api';
import { translateForm } from 'utils/FormUtils';
import { Trans } from 'react-i18next';
import { profileToEnumValue } from 'utils/ShareProfileUtils';
import { useFormatter } from 'context/FormatterContext';

const Fields: UI.FormFieldDefinition[] = [
  {
    key: 'path',
    title: 'Path',
    type: API.SettingTypeEnum.DIRECTORY_PATH,
  },
  {
    key: 'virtual_name',
    title: 'Virtual name',
    type: API.SettingTypeEnum.STRING,
    help: 'Directories with identical virtual names will be merged in filelist',
  },
  {
    key: 'profiles',
    type: API.SettingTypeEnum.LIST,
    item_type: API.SettingTypeEnum.NUMBER,
    title: 'Share profiles',
    help: 'New share profiles can be created from application settings',
  },
  {
    key: 'incoming',
    title: 'Incoming',
    type: API.SettingTypeEnum.BOOLEAN,
  },
];

export interface ShareDirectoryDialogProps {
  shareT: UI.ModuleTranslator;
}

export interface DataProps extends ShareProfileDecoratorChildProps {
  virtualNames: string[];
  rootEntry?: API.ShareRootEntryBase;
}

interface Entry extends ShareRootEntryBase, UI.FormValueMap {}

type Props = ShareDirectoryDialogProps & DataProps & ModalRouteDecoratorChildProps;

const ShareDirectoryDialog: React.FC<Props> = ({
  shareT,
  profiles,
  rootEntry,
  socket,
  virtualNames,
  location,
  ...other
}) => {
  const formatter = useFormatter();

  const fieldDefinitions = useMemo<UI.FormFieldDefinition[]>(() => {
    const defs = translateForm(Fields, shareT);

    const shareProfileDefinitions = defs.find((def) => def.key === 'profiles')!;
    Object.assign(shareProfileDefinitions, {
      options: profiles.map((p) => profileToEnumValue(p, formatter)),
      default_value: [profiles.find((profile) => profile.default)!.id],
    });

    return defs;
  }, []);
  const formRef = useRef<Form<Entry>>(null);

  const isNew = !rootEntry;

  const onFieldChanged: FormFieldChangeHandler<Entry> = (id, value, hasChanges) => {
    if (id.includes('path')) {
      const mergeFields = {
        virtual_name: !!value.path ? getLastDirectory(value.path) : undefined,
      };

      return Promise.resolve(mergeFields);
    }

    return null;
  };

  const handleSave = () => {
    return formRef.current!.save();
  };

  const onSave: FormSaveHandler<Entry> = (changedFields) => {
    if (isNew) {
      return socket.post(ShareRootConstants.ROOTS_URL, changedFields);
    }

    return socket.patch(`${ShareRootConstants.ROOTS_URL}/${rootEntry.id}`, changedFields);
  };

  const onFieldSetting: FormFieldSettingHandler<Entry> = (
    id,
    fieldOptions,
    formValue,
  ) => {
    if (id === 'path') {
      fieldOptions.disabled = !isNew;
      fieldOptions.config = {
        ...(fieldOptions.config || {}),
        historyId: FilesystemConstants.LOCATION_DOWNLOAD,
      };
    } else if (id === 'virtual_name') {
      fieldOptions.factory = t.form.Textbox;
      fieldOptions.template = AutoSuggestField;
      fieldOptions.config = {
        suggestionGetter: () => virtualNames,
      };
    }
  };

  const title = shareT.translate(isNew ? 'Add share directory' : 'Edit share directory');
  return (
    <RouteModal
      className="share-directory"
      title={title}
      onApprove={handleSave}
      closable={false}
      icon={IconConstants.FOLDER}
      {...other}
    >
      <Message
        title={shareT.translate('Hashing information')}
        icon={IconConstants.INFO}
        description={
          <Trans i18nKey={shareT.toI18nKey('hashingInformationDesc')}>
            <p>
              New files will appear in share only after they have finished hashing (the
              client has calculated checksums for them). Information about hashing
              progress will be posted to the event log.
            </p>
          </Trans>
        }
      />
      <Form<Entry>
        ref={formRef}
        fieldDefinitions={fieldDefinitions}
        onFieldChanged={onFieldChanged}
        onFieldSetting={onFieldSetting}
        onSave={onSave}
        sourceValue={rootEntry as Entry}
        location={location}
      />
    </RouteModal>
  );
};

export default ModalRouteDecorator<ShareDirectoryDialogProps>(
  ShareProfileDecorator<Omit<Props, keyof DataProps>>(ShareDirectoryDialog, false, {
    urls: {
      virtualNames: ShareConstants.GROUPED_ROOTS_GET_URL,
      rootEntry: ({ params }, socket) => {
        if (!params.directoryId) {
          return Promise.resolve(undefined);
        }

        return socket.get(`${ShareRootConstants.ROOTS_URL}/${params.directoryId}`);
      },
    },
    dataConverters: {
      virtualNames: (data: API.GroupedPath[]) => data.map((item) => item.name, []),
    },
  }),
  '/directories/:directoryId?',
);
