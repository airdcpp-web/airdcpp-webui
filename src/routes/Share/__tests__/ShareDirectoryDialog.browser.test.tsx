import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';

import { BaseRenderResult, renderDataRoutes } from '@/tests/render/test-renderers';

import * as UI from '@/types/ui';

import {
  createTestRouteModalController,
  TestRouteModalNavigateButton,
} from '@/tests/helpers/test-dialog-helpers';
import { getModuleT } from '@/utils/TranslationUtils';
import { useTranslation } from 'react-i18next';
import ShareProfileConstants from '@/constants/ShareProfileConstants';
import {
  ShareProfileDefault,
  ShareProfileEmpty,
  ShareProfilesListResponse,
} from '@/tests/mocks/api/share-profiles';

import {
  expectFieldValue,
  setInputFieldValues,
  setupUserEvent,
  setSelectFieldValues,
} from '@/tests/helpers/test-form-helpers';
import ShareRootConstants from '@/constants/ShareRootConstants';
import { MOCK_SHARE_ROOT_ID, ShareRootNormal } from '@/tests/mocks/api/share-roots';
import ShareDirectoryDialog from '../components/ShareDirectoryDialog';
import ShareConstants from '@/constants/ShareConstants';
import { ShareGetGroupedRootsResponse } from '@/tests/mocks/api/share';
import { fireEvent, waitFor } from '@testing-library/dom';
import FilesystemConstants from '@/constants/FilesystemConstants';
import { FilesystemListContentResponse } from '@/tests/mocks/api/filesystem';
import { getBrowseStorageKey } from '@/components/filebrowser/effects/useFileItemSelection';
import { saveLocalProperty } from '@/utils/BrowserUtils';
import { formatProfileNameWithSize } from '@/utils/ShareProfileUtils';
import {
  clickButton,
  waitExpectRequestToMatchSnapshot,
  waitForData,
} from '@/tests/helpers/test-helpers';
import { initCommonDataMocks } from '@/tests/mocks/mock-data-common';
import { getMockServer, MockServer } from '@/tests/mocks/mock-server';
import { VIEW_FIXED_HEIGHT } from '@/tests/render/test-containers';

import * as API from '@/types/api';

// tslint:disable:no-empty
describe('ShareDirectoryDialog', () => {
  let server: MockServer;

  beforeEach(() => {
    server =
      getMockServer(/*{
      delayMs: 0,
      loggerOptions: {
        logLevel: 'verbose',
      },
    }*/);
  });

  afterEach(() => {
    server.stop();
  });

  const getSocket = async () => {
    const commonData = await initCommonDataMocks(server);

    // Data fetch
    server.addRequestHandler(
      'GET',
      ShareProfileConstants.PROFILES_URL,
      ShareProfilesListResponse,
    );
    server.addRequestHandler(
      'GET',
      `${ShareRootConstants.ROOTS_URL}/${MOCK_SHARE_ROOT_ID}`,
      ShareRootNormal,
    );
    server.addRequestHandler(
      'GET',
      ShareConstants.GROUPED_ROOTS_GET_URL,
      ShareGetGroupedRootsResponse,
    );

    // Browse dialog
    server.addRequestHandler(
      'POST',
      FilesystemConstants.LIST_URL,
      FilesystemListContentResponse,
    );

    // Listeners
    server.addSubscriptionHandler(
      ShareProfileConstants.PROFILES_URL,
      'share_profile_added',
      undefined,
    );
    server.addSubscriptionHandler(
      ShareProfileConstants.PROFILES_URL,
      'share_profile_updated',
      undefined,
    );
    server.addSubscriptionHandler(
      ShareProfileConstants.PROFILES_URL,
      'share_profile_removed',
      undefined,
    );

    const onCreated = vi.fn();
    const onUpdated = vi.fn();

    // Save handlers
    server.addRequestHandler(
      'PATCH',
      `${ShareRootConstants.ROOTS_URL}/${MOCK_SHARE_ROOT_ID}`,
      undefined,
      onUpdated,
    );
    server.addRequestHandler('POST', ShareRootConstants.ROOTS_URL, undefined, onCreated);

    return { commonData, server, onCreated, onUpdated };
  };

  const renderDialog = async (id: number | string | null) => {
    const { commonData, server, ...other } = await getSocket();

    const ShareDirectoryDialogTest = () => {
      const { t } = useTranslation();
      const shareT = getModuleT(t, UI.Modules.SHARE);
      return (
        <>
          <TestRouteModalNavigateButton
            modalRoute={id ? `/home/directories/${id}` : '/home/directories'}
          />
          <ShareDirectoryDialog shareT={shareT} />
        </>
      );
    };

    const routes = [
      {
        path: '/home/*',
        Component: ShareDirectoryDialogTest,
      },
    ];

    const renderData = renderDataRoutes(routes, commonData, {
      routerProps: { initialEntries: ['/home'] },
      viewType: VIEW_FIXED_HEIGHT,
    });

    const modalController = createTestRouteModalController(renderData);
    return { modalController, ...commonData, ...renderData, ...other };
  };

  const setProfiles = async (
    {
      getByLabelText,
      getByRole,
      formatter,
    }: Pick<BaseRenderResult, 'getByLabelText' | 'getByRole' | 'formatter'>,
    profiles: API.ShareProfile[],
  ) => {
    await setSelectFieldValues(
      { getByLabelText, getByRole },
      {
        field: ShareProfileField,
        options: profiles.map((p) => ({
          id: p.id.toString(),
          label: formatProfileNameWithSize(p, formatter),
        })),
      },
    );
  };

  const setPath = async (
    { getByText, getByLabelText, getByRole, queryByText }: BaseRenderResult,
    path: string,
  ) => {
    saveLocalProperty(getBrowseStorageKey(FilesystemConstants.LOCATION_DOWNLOAD), path);

    // Open dialog and select the initial path
    expect(fireEvent.click(getByText('Browse'))).toBeTruthy();
    await waitForData('Loading items', queryByText);

    clickButton('Select', getByRole);

    await waitFor(() => {
      expectFieldValue({ getByLabelText }, 'Path', path);
    });
  };

  const ShareProfileField = {
    id: 'profiles',
    label: 'Share profiles',
  };

  test('should update existing', async () => {
    const userEvent = setupUserEvent();
    const renderResult = await renderDialog(MOCK_SHARE_ROOT_ID);
    const { getByText, getByLabelText, modalController, onUpdated } = renderResult;

    await modalController.openDialog();

    // Check content
    await waitFor(() => expect(getByText('Edit share directory')).toBeTruthy());

    // Edit
    await setInputFieldValues(
      { userEvent, getByLabelText },
      {
        'Virtual name': 'Updated name',
      },
    );

    await setProfiles(renderResult, [ShareProfileEmpty]);

    await modalController.closeDialogButton('Save');

    await waitExpectRequestToMatchSnapshot(onUpdated);
  }, 100000);

  test('should create new', async () => {
    const renderResult = await renderDialog(null);
    const { getByText, getByLabelText, modalController, onCreated } = renderResult;

    await modalController.openDialog();

    // Check content
    await waitFor(() => expect(getByText('Add share directory')).toBeTruthy());

    // Edit

    // Profiles
    await setProfiles(renderResult, [ShareProfileDefault, ShareProfileEmpty]);

    // Path
    await setPath(renderResult, '/home/airdcpp/Downloads/');

    // Virtual name should be set based on the path
    await waitFor(() => {
      expectFieldValue({ getByLabelText }, 'Virtual name', 'Downloads');
    });

    await modalController.closeDialogButton('Save');

    await waitExpectRequestToMatchSnapshot(onCreated);
  }, 100000);
});
