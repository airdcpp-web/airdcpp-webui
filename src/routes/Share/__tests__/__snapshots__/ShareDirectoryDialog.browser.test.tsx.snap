// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ShareDirectoryDialog > should create new 1`] = `
{
  "data": {
    "incoming": false,
    "path": "/home/airdcpp/Downloads/",
    "profiles": [
      28550432,
      265571564,
    ],
    "virtual_name": "Downloads",
  },
  "method": "POST",
  "path": "share_roots",
}
`;

exports[`ShareDirectoryDialog > should update existing 1`] = `
{
  "data": {
    "profiles": [
      265571564,
    ],
    "virtual_name": "Updated name",
  },
  "method": "PATCH",
  "path": "share_roots/YDCXPCVSPNBXH2WFTKTBZIFZBYNWRWRT2OJ5F6Y",
}
`;
